### PR TITLE
[CELEBORN-2230] SparkUtils#shouldReportShuffleFetchFailure method should retrieve the number of task failures from TaskSetManager

### PR DESCRIPTION
**Pull request by leixm**
### What changes were proposed in this pull request?
Reopen from https://github.com/apache/celeborn/pull/3556, retrieve the number of task failures from TaskSetManager in SparkUtils#shouldReportShuffleFetchFailure method

### Why are the changes needed?
https://github.com/apache/celeborn/blob/main/client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkUtils.java#L514 We record the failure counts for task attempts in the "UNKNOWN" and "FAILED" states, but spark might not record the failure counts for task attempts in the FAILED state. This is a common occurrence in our production environment where task attempts fail due to container preemption. This situation happens frequently and failure counts should not be recorded, as existing code logic makes it easier for stageRerun to be triggered prematurely. Therefore, obtaining the failure counts for task attempts from the taskSetManager would be more accurate.

### Does this PR resolve a correctness bug?

No.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

UTs.

### CODE REVIEW

**Comment by leixm** [comment]
cc @turboFei

**Comment by leixm** [comment]
cc @turboFei

**Comment by RexXiong** [comment]
Overall, this change looks good to me. The approach of retrieving failure counts directly from `TaskSetManager.numFailures` is more accurate than manually counting failed task attempts, especially for cases like container preemption where Spark doesn't increment the failure count.

One suggestion: It would be helpful to add a test case that verifies the failure count is correctly incremented after an actual task failure (e.g., simulate a task failure and then verify that `getTaskFailureCount` returns the expected increased value). Currently, the test only validates boundary conditions (initial value, out-of-bounds indices), but doesn't cover the actual failure counting scenario.

---
by claude

**Comment by leixm** [comment]
> Overall, this change looks good to me. The approach of retrieving failure counts directly from `TaskSetManager.numFailures` is more accurate than manually counting failed task attempts, especially for cases like container preemption where Spark doesn't increment the failure count.
> 
> One suggestion: It would be helpful to add a test case that verifies the failure count is correctly incremented after an actual task failure (e.g., simulate a task failure and then verify that `getTaskFailureCount` returns the expected increased value). Currently, the test only validates boundary conditions (initial value, out-of-bounds indices), but doesn't cover the actual failure counting scenario.
> 
> by claude

Added a UT, PTAL.

File: client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkUtils.java [hunk 1]
```diff
@@ -511,36 +547,32 @@ public static boolean shouldReportShuffleFetchFailure(long taskId) {
...
+        //  2. If no other taskAttempts are running, trigger a FetchFailed exception
+        //  to keep the same behavior as Spark.
+        // Note: previousFailureCount does NOT include the current failure,
+        //       so (previousFailureCount + 1) represents the total failure count.
+        int previousFailureCount = getTaskFailureCount(taskSetManager, taskInfo.index());
+        // Fail-safe: if failure count cannot be determined, conservatively trigger
+        // FetchFailed to avoid silently swallowing the error.
+        if (previousFailureCount < 0) {
+          return true;
+        }
```
[the remark this answers is not in the document]
Comment (leixm) [reply client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkUtils.java:569]: I think this is unreasonable. It shouldn't check `!hasRunningAttempt`, but should directly return `true`. Otherwise, `FetchFailed` won't be triggered and app will fail.

File: client-spark/spark-2/src/main/java/org/apache/spark/shuffle/celeborn/SparkUtils.java [hunk 2]
```diff
@@ -375,36 +411,32 @@ public static boolean shouldReportShuffleFetchFailure(long taskId) {
...
+        //  1. If total failures (previous failures + current failure) >= maxTaskFails
+        //  2. If no other taskAttempts are running, trigger a FetchFailed exception
+        //  to keep the same behavior as Spark.
+        // Note: previousFailureCount does NOT include the current failure,
+        //       so (previousFailureCount + 1) represents the total failure count.
+        int previousFailureCount = getTaskFailureCount(taskSetManager, taskInfo.index());
+        // Fail-safe: if failure count cannot be determined, conservatively trigger
+        // FetchFailed to avoid silently swallowing the error.
+        if (previousFailureCount < 0) {
+          return true;
```
[the remark this answers is not in the document]
Comment (leixm) [reply client-spark/spark-2/src/main/java/org/apache/spark/shuffle/celeborn/SparkUtils.java]: I think this is unreasonable. It shouldn't check `!hasRunningAttempt`, but should directly return `true`. Otherwise, `FetchFailed` won't be triggered and app will fail.

**Comment by RexXiong** [comment]
I think Copilot's suggestion makes sense here. When `previousFailureCount < 0` (reflection fails), checking `!hasRunningAttempt` before returning `true` would avoid premature stage reruns while still ensuring the failure is eventually reported.

Could you elaborate on why "app will fail" if we check `!hasRunningAttempt`? If there's no running attempt, the method would still return `true` and trigger `FetchFailed`. The only difference is that when another attempt is running, we defer the failure report to give that attempt a chance to succeed.

This seems consistent with the original logic where `!hasRunningAttempt` is one of the conditions to trigger `FetchFailed`.

**Comment by leixm** [comment]
> I think Copilot's suggestion makes sense here. When `previousFailureCount < 0` (reflection fails), checking `!hasRunningAttempt` before returning `true` would avoid premature stage reruns while still ensuring the failure is eventually reported.
> 
> Could you elaborate on why "app will fail" if we check `!hasRunningAttempt`? If there's no running attempt, the method would still return `true` and trigger `FetchFailed`. The only difference is that when another attempt is running, we defer the failure report to give that attempt a chance to succeed.
> 
> This seems consistent with the original logic where `!hasRunningAttempt` is one of the conditions to trigger `FetchFailed`.

I made a mistake, you're right.
